### PR TITLE
Add metrics for cache hits/misses

### PR DESCRIPTION
**Pull request by SuperQ**
Add counters for cache middleware hits and misses.
- Include test for cache hits.

### CODE REVIEW

**Comment by miekg** [comment]
Nice, could you give the dnssec middleware the same treatment? Is has a `type=signature` cache.

**Comment by SuperQ** [comment]
Sure, Would you like it in this PR or separately?

**Comment by miekg** [comment]
Think it makes sense to do it in this pr as well. Thanks!

**Comment by SuperQ** [comment]
I added it without labeling since there is only one type of cache in dnssec.  Labels can be added later if additional dnssec caches are added.

**Comment by miekg** [comment]
Looks great. But could you update the READMEs as well?

**Comment by SuperQ** [comment]
No problem, done.

File: middleware/dnssec/README.md [hunk 1]
```diff
@@ -43,5 +43,7 @@ If monitoring is enabled (via the *prometheus* directive) then the following met
 
 * coredns_dnssec_cache_size{type} - total elements in the cache, type is "signature".
```
Comment (miekg) [inline middleware/dnssec/README.md:44]: then we should remove the `type` here as well. I added it to be in sync with caching, but you have a point in removing the type from _dnssec_

Sorry for nitpicking.

File: middleware/cache/README.md [hunk 2]
```diff
@@ -39,8 +39,12 @@ The minimum TTL allowed on resource records is 5 seconds.
 
 If monitoring is enabled (via the *prometheus* directive) then the following metrics are exported:
 
-* coredns_cache_size{type} - total elements in the case, type is either "denial" or "success".
-* coredns_cache_capacity{type} - total capacity of the cache, type is either "denial" or "success".
+* coredns_cache_size{type} - Total elements in the cache by cache type.
+* coredns_cache_capacity{type} - Total capacity of the cache by cache type.
+* coredns_cache_hits_total{type} - Counter of cache hits by cache type.
+* coredns_cache_misses_total - Counter of cache misses.
```
Comment (miekg) [inline middleware/cache/README.md:45]: miss total should have a `{type}` as well.

File: middleware/dnssec/handler.go [hunk 3]
```diff
@@ -53,6 +53,20 @@ var (
 		Name:      "cache_capacity",
 		Help:      "The dnssec cache's capacity.",
 	}, []string{"type"})
```
Comment (miekg) [inline middleware/dnssec/handler.go:55]: kill the type here as well.

File: middleware/cache/README.md (hunk 2 above)
Comment (SuperQ) [reply middleware/cache/README.md:45]: Misses don't seem to have a type.  In the cache code, it checks one cache, then the other, then miss is a drop-through.

File: middleware/dnssec/handler.go (hunk 3 above)
Comment (SuperQ) [reply middleware/dnssec/handler.go:55]: Yes, I agree, dnscache only has one type, so it shouldn't be labeled.

File: middleware/dnssec/README.md (hunk 1 above)
Comment (SuperQ) [reply middleware/dnssec/README.md:44]: No, not nit-picking at all.  This is something I noticed and thought about fixing. :-)

**Comment by miekg** [comment]
[ Quoting notifications@github.com in "Re: [miekg/coredns] Add metrics for..." ]

> SuperQ commented on this pull request.
> 
> > @@ -39,8 +39,12 @@ The minimum TTL allowed on resource records is 5 seconds.
> 
> If monitoring is enabled (via the _prometheus_ directive) then the following metrics are exported:
> 
> -\* coredns_cache_size{type} - total elements in the case, type is either "denial" or "success".
> -\* coredns_cache_capacity{type} - total capacity of the cache, type is either "denial" or "success".
> +\* coredns_cache_size{type} - Total elements in the cache by cache type.
> +\* coredns_cache_capacity{type} - Total capacity of the cache by cache type.
> +\* coredns_cache_hits_total{type} - Counter of cache hits by cache type.
> +\* coredns_cache_misses_total - Counter of cache misses.
> 
> Misses don't seem to have a type.  In the cache code, it checks one cache, then the other, then miss is a drop-through.

Can you then still calculate the hit/miss percentage per cache type?

/Miek

## 

Miek Gieben

**Comment by SuperQ** [comment]
I don't think that's possible. I'm trying to think about a logic order that would allow for that in a reasonable way.

**Comment by miekg** [comment]
[ Quoting notifications@github.com in "Re: [miekg/coredns] Add metrics for..." ]

> I don't think that's possible. I'm trying to think about a logic order that would allow for that in a reasonable way.

This is when I said: screw it: just size and capacity :)

Basically you need to know which cache you are querying before hand. So we
may need to give this function a Classify/Typify and then switch on that.

**Comment by SuperQ** [comment]
You can still get a useful cache hit rate by counting the total hits from both types together.

**Comment by miekg** [comment]
[ Quoting notifications@github.com in "Re: [miekg/coredns] Add metrics for..." ]

> You can still get a useful cache hit rate by counting the total hits from both types together.

ok, let's go with that for now.

**Pull request merged**